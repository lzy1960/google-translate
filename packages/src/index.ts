import fetch from 'node-fetch'
import qs from 'qs'
import { Options } from '../types/index'

export const translate = async (
  text: string,
  options: Options = {
    from: 'auto',
    to: 'en',
    tld: 'cn',
  }
) => {
  const { from, to, tld } = options
  const url = 'https://translate.google.' + tld
  const rpcids = 'MkEWBc'
  const params = {
    rpcids,
    'source-path': '/',
    'f.sid': '6886843866985976507',
    bl: 'boq_translate-webserver_20220627.09_p0',
    hl: 'zh-CN',
    'soc-app': '1',
    'soc-platform': '1',
    'soc-device': '1',
    _reqid: '248103',
    rt: 'c',
  }

  const fullUrl =
    url + '/_/TranslateWebserverUi/data/batchexecute?' + qs.stringify(params)

  const _formData = [
    [
      [
        rpcids,
        JSON.stringify([[text, from, to, true], [null]]),
        null,
        'generic',
      ],
    ],
  ]

  const formData = new URLSearchParams({
    'f.req': JSON.stringify(_formData),
  })

  const res = await fetch(fullUrl, {
    method: 'POST',
    // headers: {
    //   'content-type': 'application/x-www-form-urlencoded;charset=UTF-8',
    // },
    body: formData,
  })
  const data = await res.text()
  // [
  //   [
  //     'wrb.fr',
  //     'MkEWBc',
  //     '[["Nǐ hǎo",null,"zh-CN",[[[0,[[[null,2]],[true]]]],2],[["你好",null,null,2]]],[[[null,null,null,null,null,[["Hello",null,null,null,[["Hello",[4,5,11]],["Hi",[4]],["Hello there",[4]]]]]]],"en",1,"zh-CN",["你好","auto","en",true]],"zh-CN",["你好!",null,null,null,null,[[["感叹词",[["Hello!",null,["你好!","喂!"],1,true],["Hi!",null,["嗨!","你好!"],1,true],["Hallo!",null,["你好!"],3,true]],"en","zh-CN"]],3],null,null,"zh-CN",1]]',
  //     null,
  //     null,
  //     null,
  //     'generic'
  //   ],
  //   [ 'di', 24 ],
  //   [ 'af.httprm', 23, '-1257359032242568457' ]
  // ]

  // 最上面的两行去掉以后
  // 总共四行
  // 第一行表示前两行的长度
  // 第三行表示后两行的长度

  // 开始处理数据
  const body = data.slice(6)
  // 先找到第一个数字
  const firstLen = /^\d+/.exec(body)![0]
  // 找到第一个数组
  const firstJson = body.slice(
    firstLen.length,
    parseInt(firstLen) + firstLen.length
  )

  const rawResult = JSON.parse(firstJson)
  const result = JSON.parse(rawResult[0][2])
  console.log(result)

  return 'Hello'
}
