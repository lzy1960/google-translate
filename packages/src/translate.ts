import fetch from 'node-fetch'
import { stringify } from 'qs'
import {
  Options,
  BatchExecute,
  ErrorCode,
  RpcIds,
  Result,
} from '../types/index'
import { Language } from '../types/language'
import { RcpIdsKeys, DefaultResult, WordResult } from '../types/index'
import { extend } from '../shared'

const DEFAULT_OPTIONS: Options = {
  from: 'auto',
  to: 'en',
  tld: 'cn',
  type: 'default',
  isMobile: false,
}
const batchKey = 'x-goog-batchexecute-bgr'
const DEFAULT_HEADERS = {
  'User-Agent':
    'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.0.0 Safari/537.36',
  [batchKey]: BatchExecute['MOBILE'],
}

export const translate = async (
  text: string,
  options: Options = DEFAULT_OPTIONS
): Promise<Result | never> => {
  const _options = extend({}, DEFAULT_OPTIONS, options)

  // 传入的from和to防错校验
  const isRightLanguage = checkFromAndTo(options)
  if (!isRightLanguage) {
    throw new Error('not support this language')
  }

  const res = await getTranslateData(text, _options)
  const data = formatBodyToRawResult(res)
  const result = getResult(data, _options)

  return result
}

export const getTranslateData = async (
  text: string,
  options: Options = DEFAULT_OPTIONS
): Promise<string | never> => {
  const _options = extend({}, DEFAULT_OPTIONS, options)
  const { from, to, tld, type } = _options
  const url = 'https://translate.google.' + tld
  const rpcids = RpcIds[type!.toUpperCase() as RcpIdsKeys]
  const params = {
    rpcids,
    'source-path': '/',
    'f.sid': '6886843866985976507',
    bl: 'boq_translate-webserver_20220627.09_p0',
    hl: 'zh-CN',
    'soc-app': '1',
    'soc-platform': '1',
    'soc-device': '1',
    _reqid: '248103',
    rt: 'c',
  }

  const fullUrl =
    url + '/_/TranslateWebserverUi/data/batchexecute?' + stringify(params)

  // 设置是否是移动端的请求头
  const _headers = { ...DEFAULT_HEADERS }
  // if (_options.isMobile) {
  //   _headers[batchKey] = BatchExecute['MOBILE']
  // } else {
  //   _headers[batchKey] = BatchExecute['PC']
  // }

  const _formData = [
    [
      [
        rpcids,
        JSON.stringify([
          [text, from, to, true],
          type === 'word' ? null : [null],
        ]),
        null,
        'generic',
      ],
    ],
  ]

  const formData = new URLSearchParams({
    'f.req': JSON.stringify(_formData),
  })

  try {
    const res = await fetch(fullUrl, {
      method: 'POST',
      body: formData,
      headers: _headers,
    })
    return res.text()
  } catch (error) {
    throw new Error(ErrorCode.BAD_REQUEST)
  }
}

export const formatBodyToRawResult = (body: string) => {
  const rawBody = body.slice(6)
  // 先找到第一个数字
  const firstLen = /^\d+/.exec(rawBody)![0]
  // 找到第一个数组
  const firstJson = rawBody.slice(
    firstLen.length,
    parseInt(firstLen) + firstLen.length
  )

  const rawResult = JSON.parse(firstJson)
  return JSON.parse(rawResult[0][2])
}

export const getResult = (data: any[] | null, options: Options): Result => {
  const _options = extend({}, DEFAULT_OPTIONS, options)
  const { from, type } = _options

  if (data) {
    switch (type) {
      case 'default':
        return processDefault(data)

      case 'word':
        return processWord(data)
    }
  }
  return {
    from,
    pronunciation: null,
    text: '',
  }
}
export const checkFromAndTo = (options: Options): boolean => {
  const { from, to } = options
  if (from in Language && to in Language) {
    return true
  } else {
    return false
  }
}

function processDefault(data: any[]): DefaultResult {
  const result = {} as DefaultResult
  const rawResultArr = data[1][0][0][5]
  if (rawResultArr) {
    result.from = data[0][2]
    result.text = rawResultArr.map((item: string[]) => item[0]).join(' ')
    result.pronunciation = data[0][0]
  }
  return result
}
function processWord(data: any[]): WordResult | never {
  if (!data.length) {
    throw new Error(ErrorCode['NO_RESULT'])
  }
  const result: WordResult = {
    text: data[0][0],
    common: [],
  }
  result.common = data[0][5][0].map((item: any) => ({
    type: item[0],
    words: item[1].map((word: any) => ({
      word: word[0],
      explains: word[2],
      frequency: word[3],
    })),
  }))
  return result
}
