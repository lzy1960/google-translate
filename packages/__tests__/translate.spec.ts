import { describe, expect, it } from 'vitest'
import { translate } from '../src'

describe('translate', () => {
  it('should fetch google api and return result', async () => {
    const result = await translate('你好')
    expect(result).toMatchSnapshot()
  })
  it.only('should translate some words', async () => {
    const result = await translate('你好')
    expect(result).toBe('Hello')
  })
})
