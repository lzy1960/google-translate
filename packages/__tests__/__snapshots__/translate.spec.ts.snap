// Vitest Snapshot v1

exports[`translate > should fetch google api and return result 1`] = `
")]}'

563
[[\\"wrb.fr\\",\\"MkEWBc\\",\\"[[\\\\\\"Nǐ hǎo\\\\\\",null,\\\\\\"zh-CN\\\\\\",[[[0,[[[null,2]],[true]]]],2],[[\\\\\\"你好\\\\\\",null,null,2]]],[[[null,null,null,null,null,[[\\\\\\"Hello\\\\\\",null,null,null,[[\\\\\\"Hello\\\\\\",[4,5,11]],[\\\\\\"Hi\\\\\\",[4]],[\\\\\\"Hello there\\\\\\",[4]]]]]]],\\\\\\"en\\\\\\",1,\\\\\\"zh-CN\\\\\\",[\\\\\\"你好\\\\\\",\\\\\\"auto\\\\\\",\\\\\\"en\\\\\\",true]],\\\\\\"zh-CN\\\\\\",[\\\\\\"你好!\\\\\\",null,null,null,null,[[[\\\\\\"感叹词\\\\\\",[[\\\\\\"Hello!\\\\\\",null,[\\\\\\"你好!\\\\\\",\\\\\\"喂!\\\\\\"],1,true],[\\\\\\"Hi!\\\\\\",null,[\\\\\\"嗨!\\\\\\",\\\\\\"你好!\\\\\\"],1,true],[\\\\\\"Hallo!\\\\\\",null,[\\\\\\"你好!\\\\\\"],3,true]],\\\\\\"en\\\\\\",\\\\\\"zh-CN\\\\\\"]],3],null,null,\\\\\\"zh-CN\\\\\\",1]]\\",null,null,null,\\"generic\\"],[\\"di\\",28],[\\"af.httprm\\",27,\\"-6713672277802631657\\"]]
25
[[\\"e\\",4,null,null,635]]
"
`;

exports[`translate > should fetch google api and return result 2`] = `
[
  [
    "wrb.fr",
    "MkEWBc",
    "[[\\"Nǐ hǎo\\",null,\\"zh-CN\\",[[[0,[[[null,2]],[true]]]],2],[[\\"你好\\",null,null,2]]],[[[null,null,null,null,null,[[\\"Hello\\",null,null,null,[[\\"Hello\\",[4,5,11]],[\\"Hi\\",[4]],[\\"Hello there\\",[4]]]]]]],\\"en\\",1,\\"zh-CN\\",[\\"你好\\",\\"auto\\",\\"en\\",true]],\\"zh-CN\\",[\\"你好!\\",null,null,null,null,[[[\\"感叹词\\",[[\\"Hello!\\",null,[\\"你好!\\",\\"喂!\\"],1,true],[\\"Hi!\\",null,[\\"嗨!\\",\\"你好!\\"],1,true],[\\"Hallo!\\",null,[\\"你好!\\"],3,true]],\\"en\\",\\"zh-CN\\"]],3],null,null,\\"zh-CN\\",1]]",
    null,
    null,
    null,
    "generic",
  ],
  [
    "di",
    23,
  ],
  [
    "af.httprm",
    23,
    "-4230975279875686577",
  ],
]
`;

exports[`translate > should translate some words 1`] = `
[
  [
    "wrb.fr",
    "MkEWBc",
    "[[\\"Nǐ hǎo\\",null,\\"zh-CN\\",[[[0,[[[null,2]],[true]]]],2],[[\\"你好\\",null,null,2]]],[[[null,null,null,null,null,[[\\"Hello\\",null,null,null,[[\\"Hello\\",[4,5,11]],[\\"Hi\\",[4]],[\\"Hello there\\",[4]]]]]]],\\"en\\",1,\\"zh-CN\\",[\\"你好\\",\\"auto\\",\\"en\\",true]],\\"zh-CN\\",[\\"你好!\\",null,null,null,null,[[[\\"感叹词\\",[[\\"Hello!\\",null,[\\"你好!\\",\\"喂!\\"],1,true],[\\"Hi!\\",null,[\\"嗨!\\",\\"你好!\\"],1,true],[\\"Hallo!\\",null,[\\"你好!\\"],3,true]],\\"en\\",\\"zh-CN\\"]],3],null,null,\\"zh-CN\\",1]]",
    null,
    null,
    null,
    "generic",
  ],
  [
    "di",
    25,
  ],
  [
    "af.httprm",
    25,
    "439813678853782960",
  ],
]
`;
